### ISSUE TITLE
[bug] Transparent window glitch on macos sonoma (14) after focus change

### ISSUE THREAD

**Issue by RoyalFoxy** [tauri-apps/tauri#8255]
### Describe the bug

I'm encountering a really strange bug which happens on any window which is transparent and a focus change occured. Images below for clarification. I couldn't confirm if the issue only happens with sonoma but seeing as sonoma is quite new and there arent any issues about this yet I thought it was clear it must be some breaking api change from apples side (as a lot of other apps have had breaking changes... I also tried testing the 2.0 alpha version which had the same effect, it works to reproduce with 1.x aswell but info was run on 2.0

Before focus change
<img width="1800" alt="Screenshot 2023-11-19 at 01 29 23" src="https://github.com/tauri-apps/tauri/assets/48829979/6c1875dd-268e-4a7d-8160-9b49ace94de8">

After focus change
<img width="1800" alt="Screenshot 2023-11-19 at 01 29 32" src="https://github.com/tauri-apps/tauri/assets/48829979/38c8e87a-06ba-4742-ab66-f9e9c4415d1b">

### Reproduction

Create a new tauri app. Set the window to transparent and enable macos private api. Go to your index.html and change the body background to something like #0000004c. Start dev or create a build and switch focus back and forth, the result from the images provided should be the same.

### Expected behavior

A normal transparent window :)

### Platform and versions

```text
[✔] Environment
    - OS: Mac OS 14.2.0 X64
    ✔ Xcode Command Line Tools: installed
    ✔ rustc: 1.75.0-nightly (c543b6f35 2023-10-14)
    ✔ cargo: 1.75.0-nightly (6fa6fdc76 2023-10-10)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: nightly-aarch64-apple-darwin (default)
    - node: 21.0.0
    - pnpm: 8.9.2
    - npm: 10.2.0
    - bun: 1.0.1
 
[-] Packages
    - tauri [RUST]: 2.0.0-alpha.17
    - tauri-build [RUST]: 2.0.0-alpha.11
    - wry [RUST]: 0.34.2
    - tao [RUST]: 0.23.0
    - @tauri-apps/api [NPM]: 2.0.0-alpha.11
    - @tauri-apps/cli [NPM]: 2.0.0-alpha.17

[-] App
    - build-type: bundle
    - CSP: unset
    - distDir: ../dist
    - devPath: http://localhost:1420/
    - framework: SolidJS
    - bundler: Vite
```

**Comment by NTablang** [comment]
Is anyone still having this issue? Encountering this on my tauri app too. For reference this is my config: (developing on macos)

```
    "windows": [
      {
        "title": "",
        "label": "main",
        "fullscreen": true,
        "resizable": false, 
        "decorations": false,
        "alwaysOnTop": false, 
        "transparent": true,
        "titleBarStyle": "Transparent"
      }
    ],
    ```

**Comment by NTablang** [comment]
Found a workaround here just in case anyone needed help https://github.com/tauri-apps/tauri/issues/4243

**Comment by RoyalFoxy** [comment]
@NTablang I had decorations turned on, are you sure you're encountering the same issue?

**Comment by NTablang** [comment]
@RoyalFoxy Yup! I had the same problem with the decoration being both on and off. When I implemented their workaround (more specifically [this](https://github.com/Hacksore/overlayed/blob/master/apps/desktop/src-tauri/src/window_custom.rs#L39-L43)) there now seems to be no problem on both cases

**Comment by RoyalFoxy** [comment]
Seems to deactivate shadows... Thats personally not what I would like :/

**Comment by NTablang** [comment]
A potential hacky way could be:
- focus change
- app in the background sets shadows to false
- user changes focus back to the app
- app in the background sets shadows to true

But I don't know if that's actually gonna work lol

**Comment by 1111mp** [comment]
This issue still exists in the latest `v2 Beta` version:
```toml
[build-dependencies]
tauri-build = { version = "2.0.0-beta", features = [] }

[dependencies]
tauri = { version = "2.0.0-beta", features = [
	"image-ico",
	"macos-private-api",
	"protocol-asset",
	"tray-icon",
	"devtools",
] }
tauri-plugin-shell = "2.0.0-beta"
```

```json
{
  // ...
  "app": {
      "macOSPrivateApi": true,
  }
}
```
```rs
let builder = tauri::WebviewWindowBuilder::new(
        app_handle,
        "main",
        tauri::WebviewUrl::App("index.html".into()),
    )
    .title("NVM-Desktop")
    .visible(false)
    .fullscreen(false)
    .inner_size(1024.0, 728.0)
    .resizable(false)
    .center();
// ...    
let window = builder
        .decorations(true)
        .transparent(true)
        .hidden_title(true)
	.shadow(true)
        .title_bar_style(tauri::TitleBarStyle::Overlay)
        .build()?;
```
This problem does not occur if `app.set_activation_policy(tauri::ActivationPolicy::Accessory)` is set.
```rs
#[cfg(target_os = "macos")]
    app.set_activation_policy(tauri::ActivationPolicy::Accessory);
```

**Comment by emipasca** [comment]
> ```rust
> #[cfg(target_os = "macos")]
>     app.set_activation_policy(tauri::ActivationPolicy::Accessory);
> ```

Worked! but what does this do exactly?

**Comment by ahkohd** [comment]
> ```rust
> ActivationPolicy::Accessory
> ```

It means you plan for the application to appear on the dock.

See https://developer.apple.com/documentation/appkit/nsapplication/activationpolicy/accessory

**Comment by ahkohd** [comment]
> > ```rust
> > #[cfg(target_os = "macos")]
> >     app.set_activation_policy(tauri::ActivationPolicy::Accessory);
> > ```
> 
> Worked! but what does this do exactly?

It means you plan for the application not to appear on the dock.

See https://developer.apple.com/documentation/appkit/nsapplication/activationpolicy/accessory

**Comment by 1111mp** [comment]
This problem should be related to the [Stage Manager](https://support.apple.com/en-ca/guide/mac-help/mchl534ba392/mac) function of macOS Sonoma .

> > ```rust
> > #[cfg(target_os = "macos")]
> >     app.set_activation_policy(tauri::ActivationPolicy::Accessory);
> > ```
> 
> Worked! but what does this do exactly?

When you set it this way, your application will not be affected by `Stage Manager` and will run on the current desktop.

**Comment by RoyalFoxy** [comment]
That is quite interesting. Would you say that the original bug would not appear if stage manager is turned off? I remember having used stage manager a lot at the beginning

**Comment by 0xDaffodi** [comment]
sos, so this bug can be solved now?

**Comment by RoyalFoxy** [comment]
@0xDaffodi Haven't looked at it ever since but I couldn't get it fixed on my machine. But I think the window loosing focus and your problem are two different ones

**Comment by 0xDaffodi** [comment]
> [@0xDaffodi](https://github.com/0xDaffodi) Haven't looked at it ever since but I couldn't get it fixed on my machine. But I think the window loosing focus and your problem are two different ones

Yeah, maybe. I already fixed my elemental outline ghosting issue yesterday. Just set shadow to false in the tauri conf file.
But, I still think such settings should not be made, although they do have an effect. But I don't understand why the bug occurred.

**Comment by ollyde** [comment]
Run this in onMount for HTML and invoke the command

```
#[tauri::command]
pub fn refresh_window(window: tauri::Window) -> Result<(), String> {
    println!("Refreshing window");
    // Force window refresh by toggling visibility and redrawing
    let current_size = window.outer_size().map_err(|e| e.to_string())?;

    // Temporarily resize to trigger redraw
    let temp_size = tauri::Size::Physical(tauri::PhysicalSize::new(
        current_size.width + 1,
        current_size.height + 1,
    ));

    window.set_size(temp_size).map_err(|e| e.to_string())?;
    window.set_size(current_size).map_err(|e| e.to_string())?;

    // Disable shadows to prevent shadow bug
    window.set_shadow(false).map_err(|e| e.to_string())?;
    window
        .set_background_color(Some(tauri::window::Color(0, 0, 0, 0)))
        .unwrap();
    window.set_decorations(false).map_err(|e| e.to_string())?;

    // Force window to redraw
    window.set_focus().map_err(|e| e.to_string())?;

    // Set shadow true after a small delay, but in background
    let handle = std::thread::spawn(move || {
        std::thread::sleep(std::time::Duration::from_millis(100));
        let _ = window.set_shadow(true);
    });
    handle.join().unwrap();

    Ok(())
}

```
